src/dotnet/ToolManifest/ToolManifestEditor.cs

```diff
@@ -70,20 +70,52 @@ public ToolManifestEditor(IFileSystem fileSystem = null, IDangerousFileDetector
 
             if (deserializedManifest.Tools == null)
             {
-                deserializedManifest.Tools = new Dictionary<string, SerializableLocalToolSinglePackage>();
+                deserializedManifest.Tools = new List<SerializableLocalToolSinglePackage>();
             }
 
             deserializedManifest.Tools.Add(
-                packageId.ToString(),
                 new SerializableLocalToolSinglePackage
-                {
+                {   PackageId = packageId.ToString(),
                     Version = nuGetVersion.ToNormalizedString(),
                     Commands = toolCommandNames.Select(c => c.Value).ToArray()
                 });
 
             _fileSystem.File.WriteAllText(to.Value, deserializedManifest.ToJson());
         }
 
+        public void Edit(
+            FilePath to,
+            PackageId packageId,
+            NuGetVersion newNuGetVersion,
+            ToolCommandName[] newToolCommandNames)
+        {
+            SerializableLocalToolsManifest deserializedManifest =
+                DeserializeLocalToolsManifest(to);
+
+            List<ToolManifestPackage> toolManifestPackages =
+                GetToolManifestPackageFromOneManifestFile(deserializedManifest, to, to.GetDirectoryPath());
+
+            var existing = toolManifestPackages.Where(t => t.PackageId.Equals(packageId)).ToArray();
+            if (existing.Any())
+            {
+                var existingPackage = existing.Single();
```

Comment:
Can there be more than one?  If not, can we combine the `Any` and `Single` calls into `SingleOrDefault`?

If so, we'll need to guard against this here otherwise this will result in an exception.

Reply:
we have guard in manifest editing layer, there can be only one package entry with the same name. As for SingleOrDefault. ToolManifestEditor is a internal API, there is no usage for finding nothing, so it should throw here

Reply:
I personally think it's cleaner to do:

```
var existingPackage =  toolManifestPackages.Where(t => t.PackageId.Equals(packageId)).SingleOrDefault();
if (existingPackage != null)
{
    ...
}
else
{
   throw new ...
}
```

Also, why the `ToArray` call here (and the other place that's similar to this code?

Reply:
That is a struct, I need to compare with default. And i think that more ugly.

for ToArray, i don't want to enumerate twice

Reply:
Ah, didn't realize it was a value type 😞.